src/adapter.ts

```diff
@@ -1,213 +1,222 @@
-        }
-
-        /**
-         * Dynamically create a CSSStyleSheet() and keep track of the adopted
-         * stylesheet index for later reference.
-         */
-        get cssStyleSheet() {
-            if (!this._cssStyleSheet) {
-                const index = document.adoptedStyleSheets.length;
-                this.classList.add(this.uuid);
-                this._cssStyleSheet = new CSSStyleSheet();
-                document.adoptedStyleSheets[index] = this._cssStyleSheet;
-                this.adoptedStyleSheetIndex = index;
-            }
-            return this._cssStyleSheet;
-        }
-
-        /**
-         * Return a selector for the this element as a class chain.
-         */
-        get objectClassSelector(): string {
-            return '&.' + this.classList.value.replace(/ /g, '.');
-        }
-
-        /**
-         * Set CSS for this element.
-         * It works like `<el style="">` with nest syntax.
-         */
-        set css(css: string) {
-            this._css = css;
-            this.cssStyleSheet.replaceSync(
-                this._class.cssProcess(`
-                    ${this.objectClassSelector} { ${ css } }
+export function AdapterMixin<TBase extends Constructor<HTMLElement>>(
+  Base: TBase
+) {
+  return class Adapter extends Base {
+    _class!: typeof Adapter; // instance class.
+    _cssStyleSheet?: CSSStyleSheet;
+    adoptedStyleSheetIndex!: number;
+    _uuid?: string; // instance id.
+    _css: string = '';
+
+    /** Tag name of this component */
+    static _tagName: string | null;
+
+    /** Styles which contain only css for this component */
+    static _styles: string[] = [];
+
+    /** CSS Process middleware, This function will be called
+     * before applying CSS to CSSStyleSheet.
+     */
+    static cssProcess(css: string): string {
+      return css;
+    }
+
+    /**
+     * Retreive styles for this component,
+     * also prevent inherit values from super class.
+     */
+    static get styles(): string[] {
+      if (this._styles === Object.getPrototypeOf(this).styles) {
+        this._styles = [];
+      }
+      return this._styles;
+    }
+
+    /** Retreive inherited styles for all super classes. */
+    static get inheritedStyles(): string[] {
+      let superClass = Object.getPrototypeOf(this);
+      let inheritedStyles: string[] = [];
+
+      while (superClass.styles !== undefined) {
+        inheritedStyles.push(...superClass.styles);
+        superClass = Object.getPrototypeOf(superClass);
+      }
+
+      return inheritedStyles;
+    }
+
+    /** Set CSS for this component */
+    static set css(css: string) {
+      this._styles = [css];
+
+      if (this.tagName) {
+        this.cssStyleSheet.replaceSync(
+          this.cssProcess(`${this.tagName} { ${this.css} }`)
+        );
+      }
+    }
+
+    /** Get CSS for this component, includes inherited styles */
+    static get css(): string {
+      return [...this.inheritedStyles, ...this.styles].join('\n');
+    }
```

Comment:
https://github.com/keenlycode/adapter/pull/26/files#diff-90f5592a8a85ef6ea1f847351c4ec2ad2f646d34f426d1e9feb2ed7411f344ccL53-L60

Instead of initiate `css` as empty string to can use spread operator for array directly and join later